core/templates/domain/objects/RatioObjectFactorySpec.ts

```
@@ -0,0 +1,139 @@
+    expect(() => {
+      ratio.fromRawInputString('1');
+    }).toThrowError(errors.INVALID_FORMAT);
+    // Invalid Ratio.
+    expect(() => {
+      ratio.fromRawInputString('1:3/2');
+    }).toThrowError(errors.NON_INTEGER_ELEMENTS);
+    expect(() => {
+      ratio.fromRawInputString('1:1/2:3/2');
+    }).toThrowError(errors.NON_INTEGER_ELEMENTS);
+    expect(() => {
+      ratio.fromRawInputString('1/2:2:3:4:5');
+    }).toThrowError(errors.NON_INTEGER_ELEMENTS);
+    expect(() => {
+      ratio.fromRawInputString('1:2.2');
+    }).toThrowError(errors.NON_INTEGER_ELEMENTS);
+    expect(() => {
+      ratio.fromRawInputString('1.2:2');
+    }).toThrowError(errors.NON_INTEGER_ELEMENTS);
+    // Invalid Colons.
+    expect(() => {
+      ratio.fromRawInputString('1::2::3');
+    }).toThrowError(errors.INVALID_COLONS);
+    expect(() => {
+      ratio.fromRawInputString('1:2::3');
+    }).toThrowError(errors.INVALID_COLONS);
+    expect(() => {
+      ratio.fromRawInputString('1:2:::3');
+    }).toThrowError(errors.INVALID_COLONS);
+    // Ratio with zero.
+    expect(() => {
+      ratio.fromRawInputString('1:0');
+    }).toThrowError(errors.INCLUDES_ZERO);
+    expect(() => {
+      ratio.fromRawInputString('0:0');
+    }).toThrowError(errors.INCLUDES_ZERO);
+    expect(() => {
+      ratio.fromRawInputString('0:1');
+    }).toThrowError(errors.INCLUDES_ZERO);
+  });
+
+  it('should convert to simplest form', () => {
+    expect(new Ratio([1, 2, 3]).convertToSimplestForm()).toEqual([1, 2, 3]);
+    expect(new Ratio([2, 4, 6]).convertToSimplestForm()).toEqual([1, 2, 3]);
+    expect(new Ratio([3, 6, 9]).convertToSimplestForm()).toEqual([1, 2, 3]);
+    expect(new Ratio([2, 3, 5]).convertToSimplestForm()).toEqual([2, 3, 5]);
+    expect(new Ratio([2, 4, 5]).convertToSimplestForm()).toEqual([2, 4, 5]);
+    expect(new Ratio([2, 0, 4]).convertToSimplestForm()).toEqual([2, 0, 4]);
+    expect(new Ratio([0, 0, 4]).convertToSimplestForm()).toEqual([0, 0, 4]);
```

Zeros are not allowed in the input, right? If so, I think these tests should throw an error.

---

no, it shouldn't throw an error because 0:2 is a valid ratio input and if we want to allow this input as valid we just need to remove the frontend validation nothing else. @seanlip your thoughts?

---

I thought we decided that we weren't accepting 0s in any of the ratios, per an earlier comment thread on this PR.

---

Sean insisted to add a check for zero input so that if in future we allow ratio with zero we can leverage that

---

Yup, I think that's still fine. But shouldn't the simplified value be [0, 0, 1] since the gcd of 0, 0 and 4 is 1?

---

I'm just returning the original value as it is in case of 0 i could change this but i think it should be returned as it is in case any of the term is zero

---

What you're proposing is not correct mathematically; the correct return value in such a case is [0, 0, 1].

---

Okay would do that

---

@seanlip i just need to clarify one thing that if terms are 2, 0, 4 in that case the simplest form would be 1, 0, 2 right ?

---

Yes.

---

I'm just asking this because `0` is taken as an invalid input especially in case of a list and if I do handle such cases I have to basically iterate over every element and check whether that number is zero or not.

---

@seanlip done